### ISSUE TITLE
Sharing collections

### ISSUE THREAD

**Issue by pjotrp** [genenetwork/genenetwork2#178]
Can we make it so that people can share a link to a collection through a URL? At the moment I think it is tied to a session. I think it would make a cool feature. Alternatively we could export a list of traits and import them again.

**Comment by zsloan** [comment]
My only concern regarding sharing through a URL is that I'm not sure what constitutes "too long" for the URL string. There might be situations where someone has hundreds, if not more, traits in a collection, and the name and data set for each trait would need to show up in the URL. If length is not an issue, I would have no problem with using GET for the collection page.

**Comment by pjotrp** [comment]
Since people can save collections - can't we refer to them by a key or hash?

**Comment by pjotrp** [comment]
Evey collection has a reference, e.g. collection_id=e69d104c-a816-4dc8-9a4d-e72cd2834200 which can be shared between users. Currently pasting a link does not work, probably because a session id is involved in fetching the traits. Do you think we can store this in a different way?

**Comment by zsloan** [comment]
Yeah, it uses the cookie to form the key used by Redis. This is because there is an AnonUser class that then has a list of collections assigned to it. When it fetches a particular collection, it uses the session id to grab the correct instance of AnonUser, uses AnonUser's key to get the list of collections from Redis, and finally checks which has the correct collection_id.

I'm not sure how to take the session id out of the picture, since it needs it to generate the Redis key in order to pull up the list of collections in the first place. The only way I can think of to get around this is to assign a Redis key to specific collections instead of lists of collections, but then users would no longer be able to view their collection lists since there would be nothing linking a specific user to a set of collections. I guess it might be possible to store the session id as an attribute of each collection and then grab all collections that match that attribute in order to get a specific user's collection list, but I'm afraid this might take a while if we end up with thousands of collections stored in Redis (since it would need to look at every single collection to check if it has a matching session id).

To sum up, I'm not sure how to get around this while still keeping the current structure where a Redis key is linked to a particular user (and their list of collections), but changing the structure such that keys are linked to specific collections seems like it would have its own set of problems (though maybe I'm wrong about how slow it would be to grab a collection list this way).

(I'm only referring to anonymous collections here, since logged-in user collections are stored in the database and I didn't write the code for them so I'd have to take a look and see how to make them shareable.)

**Comment by chfi** [comment]
Trello if still relevant @pjotrp
